usaspending_api/search/v2/views/spending_by_category_views/base_spending_by_location.py

```diff
@@ -0,0 +1,120 @@
+from abc import ABCMeta
+from enum import Enum
+from typing import List
+
+from django.db.models import QuerySet, Sum
+from elasticsearch_dsl import A, Q as ES_Q
+
+from usaspending_api.common.elasticsearch.search_wrappers import TransactionSearch
+from usaspending_api.common.helpers.api_helper import alias_response
+from usaspending_api.search.helpers.spending_by_category_helpers import (
+    fetch_country_name_from_code,
+    fetch_state_name_from_code,
+)
+from usaspending_api.search.v2.elasticsearch_helper import get_sum_aggregations
+from usaspending_api.search.v2.views.spending_by_category_views.base_spending_by_category import (
+    BaseSpendingByCategoryViewSet,
+)
+
+
+class LocationType(Enum):
+    COUNTY = "county"
+    DISTRICT = "congressional"
+    COUNTRY = "country"
+    STATE = "state"
+
+
+class BaseLocationViewSet(BaseSpendingByCategoryViewSet, metaclass=ABCMeta):
+    """Base class used by the different Location categories for Spending By Category"""
+
+    location_type: LocationType
+
+    def build_elasticsearch_search_with_aggregations(
+        self, filter_query: ES_Q, curr_partition: int, num_partitions: int, size: int
+    ) -> TransactionSearch:
+        # Create filtered Search object
+        search = TransactionSearch().filter(filter_query)
+
+        # Define all aggregations needed to build the response
+        group_by_location = A(
+            "terms",
+            field=f"pop_{self.location_type.value}_code",
+            include={"partition": curr_partition, "num_partitions": num_partitions},
+            size=size,
+        )
+
+        sum_aggregations = get_sum_aggregations("generated_pragmatic_obligation", self.pagination)
+
+        sum_as_cents = sum_aggregations["sum_as_cents"]
+        sum_as_dollars = sum_aggregations["sum_as_dollars"]
+        sum_bucket_sort = sum_aggregations["sum_bucket_sort"]
+
+        # Apply the aggregations to TransactionSearch object
+        search.aggs.bucket("group_by_location", group_by_location)
+        search.aggs["group_by_location"].metric("sum_as_cents", sum_as_cents).pipeline(
+            "sum_as_dollars", sum_as_dollars
+        ).pipeline("sum_bucket_sort", sum_bucket_sort)
+
+        return search
+
+    def build_elasticsearch_result(self, response: dict) -> List[dict]:
+        results = []
+        country_code_buckets = response.get("group_by_location", {}).get("buckets", [])
+        for bucket in country_code_buckets:
+            results.append(
+                {
+                    "amount": bucket.get("sum_as_dollars", {"value": 0})["value"],
+                    "code": bucket.get("key"),
+                    "id": None,
+                    # we're hitting the DB here because otherwise we have no guarantee that the transactions will have the country_name filled in
+                    "name": fetch_country_name_from_code(bucket.get("key")),
```

Comment:
Could you not modify the current transaction_delta_view sql to use the correct country name found in RefCountryCode instead of the Universal Transaction Matview? It would mean adding another join, but would help with performance so that we aren't having to hit the DB for every single country.

Reply:
I believe that this is the same discussion we had with the recipient hashes - it's better to hit the database for ~10 results at a time than it is to join to the RefCountryCode table. Although I suppose that the RefCountryCode table is much smaller than the recipient table, so it could be a better trade off. @tony-sappe any thoughts?

Reply:
I do recall encouraging devs to not add new columns to the materialized views for easier ETL. I don't recall discouraging additions to the "etl/delta" view. 

If I'm missing something, let's move the conv to slack

Reply:
I'm fine with it either way, that's just what I recalled. I'll add the changes to the transaction view.